### PR TITLE
A way to add rules during runtime

### PR DESCRIPTION
**Pull request by jonathonherbert**
This PR allows users to add rules to the validator as it's running via a Google Sheet. ~The sheet id is in the config.~ The relevant API parameters are injected via AWS parameter store.

It also exposes an endpoint to allow us to test the validator with a large number of rules generated with a set of permutations. Initial performance results are encouraging, with ~40k rules taking seconds to check against documents of ~300 words, but the dataset is artificial and real-world rulesets are likely to be more demanding.

It adds Gatling for testing performance -- `sbt gatling:test` will produce a performance report against a randomly generated text document. See CheckPermutationsSimulation for implementation details.

Tested end-to-end with prosemirror-typerighter!

### CODE REVIEW

File: test/resources/gatling.conf [hunk 1]
```diff
@@ -0,0 +1,127 @@
+#########################
```
Comment (mbarton) [inline test/resources/gatling.conf]: Maybe cut this down to just the settings we need? If it's idiomatic for Gatling projects to commit the whole thing no big deal

File: typerighter.cfn.yaml [hunk 2]
```diff
@@ -100,6 +100,8 @@ Resources:
           include "application"
           play.http.secret.key="${PlaySecret}"
           typerighter.ngramPath="/opt/ngram-data"
+          typerighter.sheetId="1bLCD7CrGMDAuuk76e0MTF1-DUNgmXrvTc1d5n2UMZR8"
```
Comment (mbarton) [inline typerighter.cfn.yaml]: As above, probably best not to be in an open repo

File: credentials.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "web": {
+    "client_id": "74025842873-45408g9k8aor6s53gg8lmpuuoj3ss275.apps.googleusercontent.com",
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply credentials.json]: Thanks for the spot! I am ... dumb.

File: typerighter.cfn.yaml (hunk 2 above)
Comment (jonathonherbert) [reply typerighter.cfn.yaml]: also as above 👍

File: app/controllers/ApiController.scala [hunk 4]
```diff
@@ -19,20 +22,41 @@ class ApiController (cc: ControllerComponents, lt: LanguageTool) extends Abstrac
...
+          Ok(json)
+        })
+      case Left(error) => Future.successful(BadRequest(s"Invalid request: $error"))
+    }
+  }
+
+  def refresh = Action.async { implicit request: Request[AnyContent] =>
+    for {
+      (rulesByCategory, ruleErrors) <- ruleManager.fetchByCategory()
+      errorsByCategory <- Future.sequence(rulesByCategory.map(categoryAndRules => {
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply app/controllers/ApiController.scala]: 👍 thanks for this, I was wincing as I wrote it

File: app/controllers/ApiController.scala [hunk 5]
```diff
@@ -19,20 +22,41 @@ class ApiController (cc: ControllerComponents, lt: LanguageTool) extends Abstrac
...
+        validatorPool.updateConfig(categoryAndRules._1, ValidatorConfig(categoryAndRules._2))
+      }).toList)
     } yield {
-      val results = lt.check(checkQuery.text)
-      val json = Json.obj(
-        "input" -> checkQuery.text,
-        "results" -> Json.toJson(results)
-      )
-      Ok(json)
+      val sheetId = configuration.getOptional[String]("typerighter.sheetId").getOrElse("Unknown sheet id")
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply app/controllers/ApiController.scala]: 👍

File: app/services/LanguageTool.scala [hunk 6]
```diff
@@ -2,30 +2,66 @@ package services
...
 
-    val instance = new JLanguageTool(language, cache, userConfig)
-    maybeLanguageModelDir.foreach(instance.activateLanguageModelRules)
-    new LanguageTool(instance)
+    // Disable all default rules by ... default
+    instance.getCategories().asScala.foreach((categoryData) => instance.disableCategory(categoryData._1))
+
+    // Add the rules provided in the config
+    Logger.info(s"Adding ${config.rules.size} rules to validator instance ${category}")
+    val ruleIngestionErrors = config.rules.foldLeft(List.empty[String])((acc, rule) => {
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply app/services/LanguageTool.scala]: Oh _nice_! Thanks!

File: app/controllers/ApiController.scala [hunk 7]
```diff
@@ -19,20 +22,41 @@ class ApiController (cc: ControllerComponents, lt: LanguageTool) extends Abstrac
...
+      Ok(Json.obj(
+        "sheetId" -> sheetId,
+        "categories" -> rulesByCategory.map((categoryAndRules) =>
+          Json.obj(
+            "category" -> categoryAndRules._1,
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply app/controllers/ApiController.scala]: double 👍

**Pull request merged**